### PR TITLE
Support scan in coprocessor

### PR DESCRIPTION
**Pull request by liukun4515**
Support `update`, `delete` and `scan` operation for coprocessor.

### CODE REVIEW

**Comment by siddontang** [comment]
please add a description for this PR

File: db/tikv/coprocessor.go [hunk 1]
```diff
@@ -74,17 +73,70 @@ func (db *coprocessor) Read(ctx context.Context, table string, key string, field
 	if err != nil {
 		return nil, err
 	}
-	//return db.table.DecodeValue(result.GetData(), fields)
+	//return db.table.DecolsdeValue(result.GetData(), fields)
```
Comment (siddontang) [inline db/tikv/coprocessor.go]: typo

File: db/tikv/coprocessor.go [hunk 2]
```diff
@@ -74,17 +73,70 @@ func (db *coprocessor) Read(ctx context.Context, table string, key string, field
...
+	data, err := dag.Marshal()
+	if err != nil {
+		return nil, err
+	}
+	req.Data = data
+	// send req
+	client := db.db.GetClient()
+	res := client.Send(ctx, &req, nil)
+	defer res.Close()
+	_, err = res.Next(ctx)
```
Comment (siddontang) [inline db/tikv/coprocessor.go]: Next only returns one item?

**Comment by siddontang** [comment]
PTAL @breeswish

File: db/tikv/coprocessor.go [hunk 3]
```diff
@@ -74,17 +73,76 @@ func (db *coprocessor) Read(ctx context.Context, table string, key string, field
...
+	if err != nil {
+		return nil, err
+	}
+	req.Data = data
+	// send req
+	client := db.db.GetClient()
+	res := client.Send(ctx, &req, nil)
+	defer res.Close()
+	for {
+		resSubset, err := res.Next(ctx)
```
Comment (siddontang) [inline db/tikv/coprocessor.go]: resSubset is []byte or others? I prefer putting it in the result

File: db/tikv/coprocessor.go (hunk 2 above)
Comment (liukun4515) [reply db/tikv/coprocessor.go]: Next returns marshaled [SelectResponse](https://github.com/pingcap/tipb/blob/master/proto/select.proto#L28:9) which contains some rows.
I have added the loop for calling the next until the nil is returned.

File: db/tikv/coprocessor.go (hunk 3 above)
Comment (liukun4515) [reply db/tikv/coprocessor.go]: resSubset contains the []byte which is marshaled from SelectResponse.

Comment (siddontang) [reply db/tikv/coprocessor.go]: can we ensure we get the count items?

Comment (liukun4515) [reply db/tikv/coprocessor.go]: if we must ensure the count for the scan, we should unmarshaled the resSubset.

File: pkg/util/table.go [hunk 4]
```diff
@@ -85,6 +111,13 @@ func (t *Table) GetPointRange(key string) kv.KeyRange {
 	return kv.KeyRange{StartKey: startKey, EndKey: startKey.PrefixNext()}
 }
 
+func (t *Table) GetScanRange(key string) kv.KeyRange {
+	startKey := tablecodec.EncodeRowKey(t.tableId, []byte(key))
+	nextTableId := t.tableId + 1
+	endKey := tablecodec.EncodeRowKeyWithHandle(nextTableId, math.MinInt64)
```
Comment (breeswish) [inline pkg/util/table.go]: @zz-jason PTAL, I'm not sure about the scan range.

File: db/tikv/coprocessor.go [hunk 5]
```diff
@@ -112,5 +170,18 @@ func (db *coprocessor) Insert(ctx context.Context, table string, key string, val
 }
 
 func (db *coprocessor) Delete(ctx context.Context, table string, key string) error {
-	return errors.New("unsupported type of delete in coprocessor")
+	// encode key
```
Comment (breeswish) [inline db/tikv/coprocessor.go]: Do we need to first query for existence before actually deletion, to simulate TiDB's access pattern here? @siddontang

Comment (liukun4515) [reply db/tikv/coprocessor.go]: In YCSB, there is no delete in core-workload, so the delete function will never be used.

File: db/tikv/coprocessor.go (hunk 3 above)
Comment (siddontang) [reply db/tikv/coprocessor.go]: why? e.g. here you use Next to get N items, not equal to count? maybe we should use N to make the map.

File: db/tikv/coprocessor.go (hunk 5 above)
Comment (siddontang) [reply db/tikv/coprocessor.go]: I prefer supporting it, it is easy. And we also need to use delete in shell mode.

File: db/tikv/coprocessor.go (hunk 3 above)
Comment (liukun4515) [reply db/tikv/coprocessor.go]: Each next returns the [copResultSubset](https://github.com/pingcap/tidb/blob/1a7e9ecf5201c29ebcf3c466e6864b3744e3450d/store/tikv/coprocessor.go#L500), we should unmarshal the []byte of copResultSubset to get the [SelectResponse] (https://github.com/pingcap/tipb/blob/b64a6ffa4ace41132bc10af8243086ea1024d46b/proto/select.proto#L28). Only from the SelectResponse, we can know how much items returned from this copResultSubset.

File: db/tikv/coprocessor.go [hunk 6]
```diff
@@ -112,5 +170,25 @@ func (db *coprocessor) Insert(ctx context.Context, table string, key string, val
...
+	rowKey := db.table.EncodeKey(key)
+	// do delete transaction
+	tx, err := db.db.Begin()
+	rowValue, err := tx.Get(rowKey)
```
Comment (siddontang) [inline db/tikv/coprocessor.go]: check begin error at first

Comment (siddontang) [reply db/tikv/coprocessor.go]: then use defer rollback

Comment (liukun4515) [reply db/tikv/coprocessor.go]: done

File: db/tikv/coprocessor.go [hunk 7]
```diff
@@ -74,17 +73,76 @@ func (db *coprocessor) Read(ctx context.Context, table string, key string, field
...
+	// construct the req
+	req := kv.Request{}
+	req.Concurrency = 1
+	req.Tp = kv.ReqTypeDAG
+	req.KeyRanges = []kv.KeyRange{db.table.GetScanRange(startKey)}
+	dag := db.table.BuildDAGTableScanWithLimitReq(fields, count)
+	req.StartTs = dag.StartTs
+	data, err := dag.Marshal()
+	if err != nil {
+		return nil, err
```
Comment (zz-jason) [inline db/tikv/coprocessor.go]: maybe we can `import github.com/juju/errors` and:
```go
return nil, errors.Trace(err)
```

File: db/tikv/coprocessor.go [hunk 8]
```diff
@@ -74,17 +73,76 @@ func (db *coprocessor) Read(ctx context.Context, table string, key string, field
...
 		return nil, err
 	}
+
 	//return db.table.DecodeValue(result.GetData(), fields)
 	return make(map[string][]byte, len(fields)), nil
 }
 
 func (db *coprocessor) Scan(ctx context.Context, table string, startKey string, count int, fields []string) ([]map[string][]byte, error) {
+	// construct the req
+	req := kv.Request{}
```
Comment (zz-jason) [inline db/tikv/coprocessor.go]:
```go
req := kv.Request{
    ...
}
```

File: pkg/util/table.go (hunk 4 above)
Comment (zz-jason) [reply pkg/util/table.go]: how about:
```go
endKey := tablecodec.EncodeRowKeyWithHandle(t.tableId, math.MaxInt64)
endKey = endKey.PrefixNext()
```

File: db/tikv/coprocessor.go (hunk 7 above)
Comment (siddontang) [reply db/tikv/coprocessor.go]: no need error, we don't care the root cause in this repo.

File: pkg/util/table.go (hunk 4 above)
Comment (liukun4515) [reply pkg/util/table.go]: The bytes order of the `endkey` is greater than all key in the table which we want to scan, so we can scan all data in the table from startKey.
that is make sure by the TiDB encoding function:
```
// EncodeInt appends the encoded value to slice b and returns the appended slice.
// EncodeInt guarantees that the encoded value is in ascending order for comparison.
func EncodeInt(b []byte, v int64) []byte {
	var data [8]byte
	u := EncodeIntToCmpUint(v)
	binary.BigEndian.PutUint64(data[:], u)
	return append(b, data[:]...)
}
```

**Comment by siddontang** [comment]
PTAL @zz-jason

**Comment by zz-jason** [comment]
LGTM

**Pull request merged**